core/src/main/java/com/segment/analytics/kotlin/core/Telemetry.kt

```diff
@@ -0,0 +1,271 @@
+        connection.setRequestProperty("Content-Type", "text/plain")
+        connection.doOutput = true
+        connection.setChunkedStreamingMode(0)
+        return connection
+    }
+}
+
+@Serializable
+// Default values will not be sent by Json.encodeToString
+data class RemoteMetric(
+    val type: String,
+    val metric: String,
+    var value: Int,
+    val tags: Map<String, String>,
+    val log: Map<String, String>? = null,
+)
+private const val METRIC_TYPE = "Counter"
+
+fun logError(err: Throwable) {
+    Analytics.reportInternalError(err)
+}
+
+object Telemetry {
+    // Metric class for Analytics SDK
+    const val INVOKE = "analytics_mobile.invoke"
+    // Metric class for Analytics SDK errors
+    const val INVOKE_ERROR = "analytics_mobile.invoke.error"
+    // Metric class for Analytics SDK plugins
+    const val INTEGRATION = "analytics_mobile.integration.invoke"
+    // Metric class for Analytics SDK plugin errors
+    const val INTEGRATION_ERROR = "analytics_mobile.integration.invoke.error"
+
+    var enable: Boolean = true
+    var host: String = Constants.DEFAULT_API_HOST
+    // 1.0 is 100%, will get set by Segment setting before start()
+    var sampleRate: Double = 1.0
+    var flushTimer: Int = 30 * 1000 // 30s
+    var httpClient: HTTPClient = HTTPClient("", MetricsRequestFactory())
+    var sendWriteKeyOnError: Boolean = true
+    var sendErrorLogData: Boolean = false
+    var errorHandler: ((Throwable) -> Unit)? = ::logError
+    var maxQueueSize: Int = 20
+    var errorLogSizeMax: Int = 4000
+
+    private const val MAX_QUEUE_BYTES = 28000
+    var maxQueueBytes: Int = MAX_QUEUE_BYTES
+        set(value) {
+            field = min(value, MAX_QUEUE_BYTES)
+        }
+
+    private val queue = mutableListOf<RemoteMetric>()
+    private var queueBytes = 0
+    private var queueSizeExceeded = false
+    private val seenErrors = mutableMapOf<String, Int>()
+    private var started = false
+    private var rateLimitEndTime: Long = 0
+    private var telemetryScope: CoroutineScope? = null
+    private var telemetryDispatcher: ExecutorCoroutineDispatcher? = null
+    fun start() {
+        if (started || sampleRate == 0.0) return
+        started = true
+
+        // Assume sampleRate is now set and everything in the queue hasn't had it applied
+        if (Math.random() > sampleRate) {
+            resetQueue()
+        } else {
+            queue.forEach {
+                it.value = (it.value / sampleRate).roundToInt()
+            }
+        }
+
+        val exceptionHandler = CoroutineExceptionHandler { _, t ->
+            errorHandler?.let {
+                it( Exception(
+                    "Caught Exception in Telemetry Scope: ${t.message}",
+                    t
+                ))
+            }
+        }
+        telemetryScope = CoroutineScope(SupervisorJob() + exceptionHandler)
+        telemetryDispatcher = Executors.newCachedThreadPool().asCoroutineDispatcher()
```

Comment:
I think we only need one thread for telemetry. maybe `newSingleThread` is better?